### DISCUSSION TITLE
Can I compile egui wasm to native executable?

### DISCUSSION THREAD

**Discussion by kitepad** [emilk/egui#2360] [Q&A]
Wasmer (wasmer.io) announced wasmer 3.0 and can compile to native executable. Can I use it to compile wasm that produced by egui to native executable?

**Comment by emilk** [comment]
Sure, but why compile to wasm first and then to native instead of compiling to native directly?

**Reply by coderedart** [reply]
its in the spirit of language server. all languages target lsp and all editors support lsp api. MxN combinations turns into M + N combinations. 

people write in M languages  that supports wasm. Instead of compiling to N different platforms, they only target WASI. then, WASI takes care of cross platform compilation from WASM. basically wasm being used for cross platform object (`libmyeguiapp.a`) files.

wasmer's WAPM is trying to be like flathub (flatpak) or playstore (apk), but for wasm apps. you write/upload wasm binaries, and they will cross compile to all targets without you ever having to worry about the platform toolchain stuff.

**Comment by kitepad** [comment]
Only want to one project, one project config.

**Comment by coderedart** [comment]
I don't think so.
you will need to access native APIs like windowing / graphics which are not accessible from wasm32-wasi AFAIK. 

the wasmer's thing is mostly about WASI (filesystem, io, process, threads, environment etc..). It can't do gl/vk or x11/wayland things or linking to native OS APIs. 

I am happy to be proven wrong though.
